Update Pilot to provide new destination.service.* attributes

douglas-reid opened the pull request:
This PR adds initial support to the new destination service attributes to Pilot (in the Mixer plugin).  The new attributes are:
- `destination.service.uid`
- `destination.service.name`
- `destination.service.namespace`
- `destination.service.host`

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/6014#" title="Author self-approved">douglas-reid</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[pilot/OWNERS](https://github.com/istio/istio/blob/master/pilot/OWNERS)~~ [douglas-reid]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

kyessenov commented:
I think `v1` directory is deprecated and frozen for any new development.

douglas-reid commented:
How, then, are we to update the functionality of `ServiceConfig()`, which impacts the config?

kyessenov commented:
Maybe copy the code to the new directory? I am just carrying the message that I'd get it if I sent this :)

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | 902f28d2575106b8bd803e4485e915f818e2aefb | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6014/istio-pilot-e2e/7777/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

douglas-reid commented:
@kyessenov ok. added a wrapper method to insert the new attributes.

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | af6d52c15f2dd45cde72654dab57061db8e73060 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6014/istio-pilot-e2e/7782/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | 61fa00017662ca211a89acab8111d183991484e6 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6014/istio-pilot-e2e/7796/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

kyessenov commented on pilot/pkg/networking/plugin/mixer/mixer.go:
```diff
@@ -333,3 +333,31 @@ func addStandardNodeAttributes(attr map[string]*mpb.Attributes_AttributeValue, p
+}
+
+func nameAndNamespace(serviceHostname string) (name, namespace string) {
+	if !strings.HasSuffix(serviceHostname, "svc.cluster.local") {
```
not sure if you care, but there is `DomainSuffix` variable that can be used here in place of `svc.cluster.local` to allow non-default cluster suffixes.

douglas-reid replied:
I do care. Any ideas on the best way to plumb that through appropriately?

rshriram replied:
I would just use the proxy's domain for the moment. We stopped plumbing this DomainSuffix thign everywhere. Its all FQDNs throughout the code.

douglas-reid replied:
@rshriram "for the moment" has me a bit nervous -- what is the replacement? proxy's domain also potentially includes the namespace, right?  Is that still OK to use?  Otherwise, I could (read: i was in-process of doing this) add `DomainSuffix` to the `Env`.

kyessenov commented on pilot/pkg/networking/plugin/mixer/mixer.go:
```diff
@@ -333,3 +333,31 @@ func addStandardNodeAttributes(attr map[string]*mpb.Attributes_AttributeValue, p
+
+func addDestinationServiceAttributes(attrs map[string]*mpb.Attributes_AttributeValue, destinationHostname string) {
+	svcName, svcNamespace := nameAndNamespace(destinationHostname)
+	attrs[v1.AttrDestinationService] = attrStringValue(destinationHostname)
```
I'm in favor of using literal strings for attributes. It makes it easy to search and find all occurrences.

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/6014#" title="Author self-approved">douglas-reid</a>*, *<a href="https://github.com/istio/istio/pull/6014#pullrequestreview-126085220" title="LGTM">kyessenov</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[pilot/OWNERS](https://github.com/istio/istio/blob/master/pilot/OWNERS)~~ [douglas-reid,kyessenov]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

kyessenov commented:
/hold
It's marked as WIP.

rshriram commented on pilot/pkg/networking/plugin/mixer/mixer.go:
```diff
@@ -333,3 +333,31 @@ func addStandardNodeAttributes(attr map[string]*mpb.Attributes_AttributeValue, p
 }
+
+func serviceConfig(serviceHostname string, dest *model.ServiceInstance, config model.IstioConfigStore, disableCheck, disableReport bool) *mccpb.ServiceConfig {
+	sc := v1.ServiceConfig(serviceHostname, dest, config, disableCheck, disableReport)
```
can you copy over the functions from that directory? I don't care about code duplication. We are going to be deleting that directory sometime or the other. So, lets not create dependencies

douglas-reid replied:
I can, but I didn't create any dependencies. This simply wrapped the existing dependency. FWIW, it is used elsewhere in this same file directly.  Thoughts?

douglas-reid replied:
went ahead and copied it over.

rshriram replied:
if you could get rid of those as well, it would be great. I am trying to avoid referring to anything in the v1 package

rshriram commented on pilot/pkg/proxy/envoy/v1/mixer.go:
```diff
@@ -295,12 +295,12 @@ func StandardNodeAttributes(prefix string, IPAddress string, ID string, labels m
 			Attributes: map[string]*mpb.Attributes_AttributeValue{
 				AttrDestinationService: {
-					Value: &mpb.Attributes_AttributeValue_StringValue{StringValue: serviceName},
+					Value: &mpb.Attributes_AttributeValue_StringValue{StringValue: serviceHostname},
```
please copy this function to the mixer plugins code

douglas-reid replied:
ok. done.

istio-testing commented:
New changes are detected. LGTM label has been removed.

douglas-reid commented:
Thanks for the quick and responsive feedback. I believe I've addressed the concerns raised so far. PTAL.

I've tested manually with bookinfo on kube, and saw the following example Mixer logs:
```
$ ki logs <istio-policy-pod> mixer -f | grep destination.service

destination.service           : productpage.default.svc.cluster.local
destination.service.host      : productpage.default.svc.cluster.local
destination.service.name      : productpage
destination.service.namespace : default
destination.service.uid       : istio://default/services/productpage
destination.serviceAccount    : default
destination.service           : reviews.default.svc.cluster.local
destination.service.host      : reviews.default.svc.cluster.local
destination.service.name      : reviews
destination.service.namespace : default
destination.service.uid       : istio://default/services/reviews
destination.serviceAccount    : default
destination.service           : reviews.default.svc.cluster.local
destination.service.host      : reviews.default.svc.cluster.local
destination.service.name      : reviews
destination.service.namespace : default
destination.service.uid       : istio://default/services/reviews
destination.serviceAccount    : default
destination.service           : details.default.svc.cluster.local
destination.service.host      : details.default.svc.cluster.local
destination.service.name      : details
destination.service.namespace : default
destination.service.uid       : istio://default/services/details
destination.serviceAccount    : default
```

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-unit-tests.sh | 68c2cb6ca1f30616d74013536edf48f0ec7966a1 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6014/istio-unit-tests/8272/) | `/test istio-unit-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

kyessenov commented:
/hold cancel

istio-testing commented:
New changes are detected. LGTM label has been removed.

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | 74653047e8f54dd56868be9aa1a5aee55d60e4b4 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6014/istio-pilot-e2e/7840/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

kyessenov commented:
/approve
/lgtm

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/6014#" title="Author self-approved">douglas-reid</a>*, *<a href="https://github.com/istio/istio/pull/6014#issuecomment-395181228" title="LGTM">kyessenov</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[pilot/OWNERS](https://github.com/istio/istio/blob/master/pilot/OWNERS)~~ [douglas-reid,kyessenov]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | 74653047e8f54dd56868be9aa1a5aee55d60e4b4 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6014/istio-pilot-e2e/7840/) | `/test istio-pilot-e2e`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 74653047e8f54dd56868be9aa1a5aee55d60e4b4 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6014/istio-pilot-e2e-envoyv2-v1alpha3/2284/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 74653047e8f54dd56868be9aa1a5aee55d60e4b4 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6014/istio-pilot-e2e-envoyv2-v1alpha3/2284/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`
prow/istio-pilot-e2e.sh | 74653047e8f54dd56868be9aa1a5aee55d60e4b4 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6014/istio-pilot-e2e/7848/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

douglas-reid commented:
/test istio-pilot-e2e-envoyv2-v1alpha3

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | 74653047e8f54dd56868be9aa1a5aee55d60e4b4 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6014/istio-pilot-e2e/7848/) | `/test istio-pilot-e2e`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 74653047e8f54dd56868be9aa1a5aee55d60e4b4 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6014/istio-pilot-e2e-envoyv2-v1alpha3/2288/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

douglas-reid commented:
/test istio-pilot-e2e-envoyv2-v1alpha3

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | 74653047e8f54dd56868be9aa1a5aee55d60e4b4 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6014/istio-pilot-e2e/7848/) | `/test istio-pilot-e2e`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 74653047e8f54dd56868be9aa1a5aee55d60e4b4 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6014/istio-pilot-e2e-envoyv2-v1alpha3/2291/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

kyessenov commented:
/test istio-pilot-e2e-envoyv2-v1alpha3

istio-testing commented:
@douglas-reid: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | 74653047e8f54dd56868be9aa1a5aee55d60e4b4 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6014/istio-pilot-e2e/7848/) | `/test istio-pilot-e2e`
prow/istio-pilot-e2e-envoyv2-v1alpha3.sh | 74653047e8f54dd56868be9aa1a5aee55d60e4b4 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6014/istio-pilot-e2e-envoyv2-v1alpha3/2295/) | `/test istio-pilot-e2e-envoyv2-v1alpha3`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | 74653047e8f54dd56868be9aa1a5aee55d60e4b4 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6014/istio-pilot-e2e/7870/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->
